### PR TITLE
Improve the QUIC_RSTREAM implementation

### PR DESCRIPTION
**Pull request by t8m**
Add API calls to avoid copying data when reading
These are ossl_quic_rstream_get_record() and
ossl_quic_rstream_release_record().

Add side storage for the stream frame data.
When there are too many packets referenced by the
receiving stream the function ossl_quic_rstream_move_to_rbuf() can be called to move the data to a ring buffer.

##### Checklist

- [ ] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
This is a draft as tests need to be added to cover the new API calls.

File: ssl/quic/quic_rstream.c
Comment (hlandau) [inline ssl/quic/quic_rstream.c:226]: It might be better to just fail here, since this should be a caller error.

**Review by hlandau — APPROVED** [review]
Looks good. One comment, decision is up to you.

File: ssl/quic/quic_rstream.c
Comment (t8m) [reply ssl/quic/quic_rstream.c:226]: If you look at the description I thought it might be convenient to call the function with SIZE_MAX to release the whole record.

Comment (hlandau) [reply ssl/quic/quic_rstream.c:226]: Hmm, okay. Could we handle that separately as an explicit sentinel value in the code? If the caller passes some wrong value which is not `SIZE_MAX` I think it should still fail.

Comment (t8m) [reply ssl/quic/quic_rstream.c:226]: OK, good idea.

Comment (t8m) [reply ssl/quic/quic_rstream.c:227]: Done

**Comment by t8m** [comment]
@hlandau still OK?

**Review by hlandau — APPROVED** [review]

**Comment by hlandau** [comment]
Needs rebase and second approval

**Comment by t8m** [comment]
Rebased. @hlandau please reconfirm.

Ping @openssl/committers for second approval.

The improvements should be used by the code, but that is a separate work task.

**Review by hlandau — APPROVED** [review]

**Comment by t8m** [comment]
One more fixup was needed.

**Review by hlandau — APPROVED** [review]

**Review by paulidale — COMMENTED** [review]
What is the reason for using a cyclic byte buffer rather than a list of datagrams?
It seems like it means more fiddling around.

A thread safe list would be beneficial here :grin:

**Comment by t8m** [comment]
> What is the reason for using a cyclic byte buffer rather than a list of datagrams? It seems like it means more fiddling around.
> 
> A thread safe list would be beneficial here grin

The reason is to avoid unnecessary allocations/deallocations.

**Comment by paulidale** [comment]
Hmmm, feels a bit like premature optimisation.

The downside will be extra copies won't it?

**Comment by t8m** [comment]
> Hmmm, feels a bit like premature optimisation.
> 
> The downside will be extra copies won't it?

Which extra copies? The ring buffer is used only when we are asked to deallocate the packets. We primarily use a linked list of received frames (with pointers to the packets holding the frames to deallocate later). The ring buffer is used only when the ossl_quic_rstream_move_to_rbuf() is called. That would be called by quic channel if there are too many packets being in process. Detecting that condition is something that still needs to be added to the quic channel implementation.

File: include/internal/ring_buf.h
Comment (mattcaswell) [inline include/internal/ring_buf.h:148]: Did you mean "read size" here?

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: ssl/quic/quic_rstream.c
Comment (mattcaswell) [inline ssl/quic/quic_rstream.c:198]: Is this actually an error? Doesn't this just mean there is no data to be read? I wonder if this should be reported as "we successfully read 0 bytes" - we need to be able to distinguish this from "real" failures such as from the ossl_assert below.

File: ssl/quic/quic_rstream.c
Comment (mattcaswell) [inline ssl/quic/quic_rstream.c:214]: The "if" above suggests "rec_len_" could be 0 here - which means we return 1 on 0 length here, but 0 on 0 length above.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: include/internal/ring_buf.h
Comment (t8m) [reply include/internal/ring_buf.h:148]: read side

File: ssl/quic/quic_rstream.c
Comment (t8m) [reply ssl/quic/quic_rstream.c:215]: Yep, I changed this to be consistent.

**Comment by t8m** [comment]
fixup pushed addressing @mattcaswell 's comments. @hlandau @mattcaswell please approve

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
CI failure looks relevant.

**Comment by t8m** [comment]
> CI failure looks relevant.

Should be fixed now.

**Comment by t8m** [comment]
@mattcaswell @hlandau please re-approve

**Comment by t8m** [comment]
Merged to master branch. Thank you for the reviews.

**Pull request closed without merging**
